### PR TITLE
feat: add YAML support to server

### PR DESCRIPTION
**Pull request by ammario**
This re-releases support links, amongst other benefits to operability.
- [x] Update cli docs generator to include YAML in table

Follow-up:
- [ ] Restore supportLinks docs
- [ ] Write docs on YAML
- [ ] Convert OptionSet to `[]*Option` to avoid all the complex indirection seen in this PR

### CODE REVIEW

**Comment by ammario** [comment]
@mafredri — this is probably 1k lines excluding generated code. I've manually tested that it works, but I'll write YAML docs and restore the support links docs in a separate PR with @bpmct to review.

I also agree with making OptionSet a `[]*Option`, which I'll do separately.

File: cli/clibase/yaml.go [hunk 1]
```diff
@@ -36,27 +44,43 @@ func deepMapNode(n *yaml.Node, path []string, headComment string) *yaml.Node {
...
+		}
+		comment := wordwrap.WrapString(
+			fmt.Sprintf("%s (default: %s, type: %s)", opt.Description, defValue, opt.Value.Type()),
+			80,
+		)
 		nameNode := yaml.Node{
 			Kind:        yaml.ScalarNode,
 			Value:       opt.YAML,
-			HeadComment: wordwrap.WrapString(opt.Description, 80),
+			HeadComment: wordwrap.WrapString(comment, 80),
```
Comment (coadler) [inline cli/clibase/yaml.go:75]: `comment` seems to be wordwrap'd twice

File: cli/clibase/yaml.go [hunk 2]
```diff
@@ -36,27 +44,43 @@ func deepMapNode(n *yaml.Node, path []string, headComment string) *yaml.Node {
...
 		if opt.YAML == "" {
 			continue
 		}
+
+		defValue := opt.Default
+		if defValue == "" {
+			defValue = "<unset>"
+		}
+		comment := wordwrap.WrapString(
+			fmt.Sprintf("%s (default: %s, type: %s)", opt.Description, defValue, opt.Value.Type()),
```
Comment (coadler) [inline cli/clibase/yaml.go:69]: I think this would be nicer to read as `%s\n(default: %s, type: %s)` so default is always on the last line. Looking at the golden file it was hard to parse when the default/type randomly wrapped at 80 chars.

Comment (coadler) [reply cli/clibase/yaml.go:69]: Big fan of including the description as a comment, and the word wrapping makes it easy to read.

**Review by coadler — APPROVED** [review]

File: cli/clibase/yaml.go (hunk 2 above)
Comment (ammario) [reply cli/clibase/yaml.go]: Good catch

File: cli/testdata/server-config.yaml.golden [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        regionName: Coder Embedded Relay
+        # Addresses for STUN servers to establish P2P connections. Set empty to disable
+        # P2P connections.
+        # (default: stun.l.google.com:19302, type: string-array)
+        stunAddresses:
+            - stun.l.google.com:19302
+        # An HTTP URL that is accessible by other replicas to relay DERP traffic. Required
+        # for high availability.
+        # (default: <unset>, type: url)
+        relayURL:
```
Comment (coadler) [inline cli/testdata/server-config.yaml.golden:116]: It seems like url types are the only fields that are printed out empty when the default is unset, the others seem to use the zero value.

Comment (coadler) [reply cli/testdata/server-config.yaml.golden:116]: Maybe the type would be better as `string(url)` so that it's clear you're inputting a string here? Unsure though.

File: cli/testdata/server-config.yaml.golden [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+    # Specifies the wildcard hostname to use for workspace applications in the form
+    # "*.example.com".
+    # (default: <unset>, type: url)
+    wildcardAccessURL:
+    # Specifies whether to redirect requests that do not match the access URL host.
+    # (default: <unset>, type: bool)
+    redirectToAccessURL: false
+    http:
+        # HTTP bind address of the server. Unset to disable the HTTP endpoint.
+        # (default: 127.0.0.1:3000, type: string)
```
Comment (coadler) [inline cli/testdata/server-config.yaml.golden:14]: small nit: I think standard YAML indentation is 2 spaces. Whenever I use `--write-config` my editor just changes all the indentation when saving.

Comment (ammario) [reply cli/testdata/server-config.yaml.golden:14]: Good catch

File: cli/testdata/server-config.yaml.golden (hunk 3 above)
Comment (ammario) [reply cli/testdata/server-config.yaml.golden]: I think for the most part it's self-explanatory.
